streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -182,6 +199,120 @@ private boolean lockStateDirectory() {
+                        streamsMetrics,
+                        dummyCache
+                    );
+
+                    final Task task = new StandbyTask(
+                        id,
+                        inputPartitions,
+                        topology,
+                        topologyMetadata.taskConfig(id),
+                        streamsMetrics,
+                        stateManager,
+                        this,
+                        dummyCache,
+                        context
+                    );
+
+                    // initialize and suspend new Tasks
+                    try {
+                        task.initializeIfNeeded();
+                        task.suspend();
+
+                        // add new Tasks to tasksForLocalState
+                        tasksForLocalState.put(id, task);
+                    } catch (final TaskCorruptedException e) {
+                        // Task is corrupt - wipe it out (under EOS) and don't initialize a Standby for it
+                        task.suspend();
+                        task.closeDirty();
+                    }
+                }
+            }
+        }
+    }
+
+    public boolean hasInitialTasks() {
+        return !tasksForLocalState.isEmpty();
+    }
+
+    public Task assignInitialTask(final TaskId taskId) {
+        final Task task = tasksForLocalState.remove(taskId);
+        if (task != null) {
+            lockedTasksToOwner.replace(taskId, Thread.currentThread());
+        }
+        return task;
+    }
+
+    public void closeInitialTasksIfLastAssginedThread() {
+        if (hasInitialTasks() && threadsWithAssignment.incrementAndGet() >= config.getInt(StreamsConfig.NUM_STREAM_THREADS_CONFIG)) {
+            // we need to be careful here, because other StreamThreads may still be assigning tasks (via assignInitialTask)
+            // so we first "drain" our Map of all remaining Tasks, and then close all the Tasks we successfully claimed from the Map
+            final Set<Task> tasksToClose = drainInitialTasks();
+            for (final Task task : tasksToClose) {
+                task.closeClean();
+            }
+        }
+    }
+
+    private void closeRemainingInitialTasks() {
+        closeRemainingInitialTasks(t -> true);
+    }
+
+    private void closeRemainingInitialTasks(final Predicate<Task> predicate) {
+        final Set<Task> drainedTasks = drainInitialTasks(predicate);
+        for (final Task task : drainedTasks) {
+            task.closeClean();
+        }
+    }
+
+    private Set<Task> drainInitialTasks() {
+        return drainInitialTasks(t -> true);
+    }
+
+    private Set<Task> drainInitialTasks(final Predicate<Task> predicate) {
+        final Set<Task> drainedTasks = new HashSet<>(tasksForLocalState.size());
+        for (final Map.Entry<TaskId, Task> entry : tasksForLocalState.entrySet()) {
+            if (predicate.test(entry.getValue()) && tasksForLocalState.remove(entry.getKey()) != null) {
```

Comment:
How could it be `null` ?

Reply:
This method can be called from multiple threads, during `close()`.

They can be iterating and removing elements from `tasksForLocalState` at the same time, so we therefore need to ensure that this is done atomically, and that each thread only "drains" the Task that it removed from the collection, and not any Tasks that another thread removed first.

This is safe, because `tasksForLocalState` is a `ConcurrentHashMap`, which guarantees atomicity of `remove` and iteration of `entrySet`.

Reply:
Is this on the hot code path? It seems not? -- I understand that you like to avoid `synchronized` but to me it's also a question about code which is easy to maintain and robust, and this might be a per-mature optimization?

There is the danger that somebody removes this condition in the future, not understanding the impact...

If you insist on this, let's at least add a comment why this check is necessary and should not be removed.

Reply:
While this specific method is not on the hot code path, `synchronized` methods lock the whole object, so any other `syncrhonized` method that _is_ on the hot code path, could be affected. This is why it's generally good practice to avoid `synchronized` methods, and instead use either concurrent data structures (as here), or a more fine-grained lock (from `java.util.concurrent.locks`) to lock only the resource that needs to be locked.

I'll add a comment to explain what's going on here.

Reply:
Done